### PR TITLE
fix(gateway): route URL prefix rewriting

### PR DESCRIPTION
**Pull request by michaelbeaumont**
In the presence of us creating multiple Envoy matches from one `MeshGatewayRoute`, rewriting didn't work in every case. 

### Checklist prior to review

- [x] Link to docs PR or issue --
- [x] Link to UI issue or PR --
- [x] Is the [issue worked on linked][1]? -- none
- [x] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [x] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Unit Tests --
- [x] E2E Tests --
- [x] Manual Universal Tests --
- [x] Manual Kubernetes Tests --
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [x] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch)?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/resources/apis/mesh/gateway_route_validator_test.go [hunk 1]
```diff
@@ -649,6 +649,31 @@ conf:
         - weight: 5
           destination:
             kuma.io/service: target-2
+`),
+		ErrorCases("prefix with trailing slash", []validators.Violation{{
```
Comment (slonka) [inline pkg/core/resources/apis/mesh/gateway_route_validator_test.go:653]: we have a case with a prefix and a suffix, don't we need two test for these?

File: pkg/plugins/runtime/gateway/gateway_route_generator.go [hunk 2]
```diff
@@ -95,6 +95,10 @@ func GenerateEnvoyRouteEntries(host GatewayHost) []route.Entry {
 			// Make sure the prefix has a trailing '/' so that it only matches
 			// complete path components.
 			e.Match.PrefixPath = exactPath + "/"
+			if rw := e.Rewrite; rw != nil && rw.ReplacePrefixMatch != nil {
+				prefix := strings.TrimRight(*rw.ReplacePrefixMatch, "/") + "/"
+				rw.ReplacePrefixMatch = &prefix
```
Comment (slonka) [inline pkg/plugins/runtime/gateway/gateway_route_generator.go:100]: Why did we call that `ReplacePrefixMatch`? Envoy calls it `PrefixRewrite`, it seems that it's just an unnecessary mapping I have to do in my head.

File: pkg/plugins/runtime/gateway/testdata/https/drop-prefix-gateway-route.yaml [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        responseHeadersToAdd:
+        - append: false
+          header:
+            key: Strict-Transport-Security
+            value: max-age=31536000; includeSubDomains
+        routes:
+        - match:
+            path: /prefix/a
+          route:
+            regexRewrite:
```
Comment (slonka) [inline pkg/plugins/runtime/gateway/testdata/https/drop-prefix-gateway-route.yaml:149]: why do we need regex here, woudldn't two prefix matches like in the docs https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-prefix-rewrite be enough?

**Review by slonka — COMMENTED** [review]
I'm not sure I 100% understand this, asked a couple of questions.

File: pkg/plugins/runtime/gateway/gateway_route_generator.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/gateway/gateway_route_generator.go:100]: [Gateway API](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPPathModifier)

We aren't _just_ an Envoy wrapper :shrug:

File: pkg/core/resources/apis/mesh/gateway_route_validator_test.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/core/resources/apis/mesh/gateway_route_validator_test.go:653]: This case tests that, that's why we get two errors. Is that what you mean? I'll make the name clearer.

File: pkg/plugins/runtime/gateway/gateway_route_generator.go (hunk 2 above)
Comment (slonka) [reply pkg/plugins/runtime/gateway/gateway_route_generator.go:100]: Oh, if it's in Gateway API then 👍 - I just thought that we had a different name for the sake of having a different name. Thanks for clarifying! 

> just an Envoy wrapper

I know but if there is a 1:1 relation with some fields I think it's ok to re-use the name to make it easier to read, WDYT?

File: pkg/core/resources/apis/mesh/gateway_route_validator_test.go (hunk 1 above)
Comment (slonka) [reply pkg/core/resources/apis/mesh/gateway_route_validator_test.go:653]: Did not notice that we have 2 errors, if you can adjust the name that would be great.

File: pkg/plugins/runtime/gateway/testdata/https/drop-prefix-gateway-route.yaml (hunk 3 above)
Comment (slonka) [reply pkg/plugins/runtime/gateway/testdata/https/drop-prefix-gateway-route.yaml:149]: If this is how it's "generally" done then 👍 - no worries from me.

**Review by slonka — APPROVED** [review]

File: pkg/plugins/runtime/gateway/gateway_route_generator.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/gateway/gateway_route_generator.go:100]: > I know but if there is a 1:1 relation with some fields I think it's ok to re-use the name to make it easier to read, WDYT?

Agreed in general but this is the case I think where the relation isn't 1-1, here we're still in the domain of Gateway API/MeshGatewayRoute matchers and transforming MeshGatewayRoute matchers into other MeshGatewayRoute matchers, the transformation to Envoy config happens later.

**Pull request merged**